Add hierarchy to GameObject class

MatthewWhite opened the pull request:
- Made GameObject class hierarchical
- Overhauled GameObjectManager to load hierarchies of GameObjects

MatthewWhite committed:
Add hierarchy to GameObject class

* Made GameObject class hierarchical
* Overhauled GameObjectManager to load hierarchies of GameObjects

MatthewWhite commented on HeatStroke/GOComponents/GameObject.cpp:
```diff
@@ -113,6 +120,66 @@ void GameObject::DeleteAllComponents()
+	{
+		m_mChildMap.insert(std::pair<std::string, GameObject*>(p_pChild->GetGUID(), p_pChild));
+		p_pChild->SetParent(this);
+		m_Transform.AddChild(&(p_pChild->GetTransform()));
```
@David-Hanna The hierarchical transform doesn't seem to be working properly.  Not sure if it's an unknown bug with `HierarchicalTransform` or if I'm doing something wrong.

Running the game and checking the console will show the error.  Essentially, it seems like the children are just getting the transform of their parent, not a combination of their own and their parent.  Translate seems to be fixed if you call `SetTranslate(x, y, z)` on the parent.  This doesn't seem to work for scale.

David-Hanna replied:
It looks like your `SetParent()` method is the problem. `SetParent()` calls `RemoveChild()` on the previous parent, which then calls `SetParent()` again, which calls `RemoveChild()` again, etc., until `m_pParent` is definitely a `nullptr`. I had mentioned you watch out for this. You will have to resolve it before it can be accepted.

MatthewWhite replied:
The call to `GameObject::RemoveChild` inside of `GameObject::SetParent` isn't being called, as `pOldParent == nullptr`, so I'm not sure if that's the problem.  I may be missing something though.

David-Hanna replied:
Okay, well I'd need to sit down and have some debug prints and such to see this error through any further. For now I just wanted to point out the concern between those methods calling each other back and forth. I also see now that `SetParent()` is private. Have you checked out how it's done in HierarchicalTransform? It avoids methods calling each other and allows `SetParent()` to be public.
Edit: It's also worth pointing out on this note that part of how it's done relies on the idea that in C++ classes are always their own `friend`.

MatthewWhite committed:
Fix issue where passing nullptr to HeatStroke::Transform::ParseTransformNode() threw an exception

This could occasionally cause issues with the hierarchical game objects, so I figured it would be appropriate to include it in thise feature

MatthewWhite committed:
Merge branch 'develop' into feature/hierarchical-game-object

# Conflicts:
#	HeatStroke/GOComponents/GameObject.cpp
#	HeatStroke/GOComponents/GameObject.h

MatthewWhite committed:
Fixed bug with hierarchical transform

Redefining scale, rotation, and translation components in HierarchicalTransform resulted in components in Transform being given proper values, while components in HierarchicalTransform were not.  Removing definition of components from HierarchicalTransform and relying on inheritance solves the problem.

MatthewWhite committed:
Merge branch 'develop' into feature/hierarchical-game-object

# Conflicts:
#	CS483/CS483/Kartaclysm/Data/test_level.xml
#	CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateRacing.cpp
#	HeatStroke/Services/GameObjectManager/GameObjectManager.cpp

MatthewWhite committed:
Merge branch 'develop' into feature/hierarchical-game-object

# Conflicts:
#	CS483/CS483/Kartaclysm/Data/camera.xml
#	CS483/CS483/Kartaclysm/Data/kart.xml
#	CS483/CS483/Kartaclysm/Data/test_level.xml
#	CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateRacing.cpp
#	HeatStroke/Services/GameObjectManager/GameObjectManager.cpp

MatthewWhite committed:
Tidy up GameObject::AddChild and GameObject::SetParent as per PR

MatthewWhite committed:
Fix issue where parent game object would hold on to reference of deleted children
